[PLATFORM-1071] Docs refactor

fonty1 opened the pull request:
PR includes:
- Each section is now it's own page
- Refactored docs navigation to support the above
- New text content and responsive images
- Module reference pages now split into categories
- New style guide implementation

PR doesn't include
- Planned updates to Module Reference page - intro text & canvas module rendering

Abstract project: `Streamr Docs` for content and style guide.

PR is yuge but it's mostly raw text content which doesn't require review by other devs.

timoxley commented on app/src/docs/components/DocsLayout/Navigation/index.jsx:
```diff
@@ -54,55 +63,69 @@ class Navigation extends React.Component<Props, State> {
+            }
+        })
+
+        return match
```
minor, but `some` stops iterating as soon as it finds a match, while `forEach` always iterates over the whole list:

```js
isActiveSection(subNavList) {
    return Object.keys(subNavList).some((subKey) => (
        this.props.location.pathname.includes(subNavList[subKey])
    ))
}
```

fonty1 replied:
this makes me very happy : ) thanks for the tip

timoxley commented on app/src/docs/components/DocsLayout/Navigation/index.jsx:
```diff
@@ -54,55 +63,69 @@ class Navigation extends React.Component<Props, State> {
+                    })}
+                >
+                    {/* Render subNav contents */}
+                    {Object.keys(navItem).map((subKey) => {
```
A `.filter` first would reduce 3 levels of indentation for the `<li>` to 1:

```js
Object.keys(navItem).filter((subKey) => subKey !== 'root').map((subKey) => (
    <li key={subKey} className={styles.navListItem}>
        …
    </li>
))
```

timoxley commented on app/src/docs/components/DocsLayout/Navigation/index.jsx:
```diff
@@ -122,8 +145,14 @@ class Navigation extends React.Component<Props, State> {
                 >
-                    {!!responsive && this.parseCurrentPage()}
-                    {this.parseNavigation()}
+                    {responsive && (
```
`!!responsive`, if `responsive` is ever `0` or `NaN` then it'll render those as text. Low cost, guarantees no edge case surprises.

timoxley commented on app/src/docs/components/DocsLayout/Navigation/index.jsx:
```diff
@@ -122,8 +145,14 @@ class Navigation extends React.Component<Props, State> {
+                            </Link>
+                        </li>
+                    )}
+                    {Object.keys(docsNav).map((topLevelNav, index) => this.generateNavigation(topLevelNav, index))}
```
if it were me, I'd make this a new component, rather than a method. Low priority.

timoxley commented on app/src/docs/components/DocsPages/API/UsingStreamsViaApi/index.jsx:
```diff
@@ -0,0 +1,17 @@
+
+const UsingStreamsViaApi = () => (
+    <DocsLayout>
+        <Helmet title="Work with streams via API | Streamr Docs" />
```
Did you know about `titleTemplate`? https://github.com/nfl/react-helmet#reference-guide

Maybe could centralise this like:

```js
import React from 'react'
import { Helmet } from 'react-helmet'

export default function DocsHelmet({ titleTemplate = '% | Streamr Docs', ...props }) {
    return <Helmet titleTemplate={titleTemplate} {...props} />
}
```

Then it'd only need a single file change to change how we organise the title.

fonty1 replied:
That's not a bad idea! 👍

timoxley commented on app/src/docs/components/DocsPages/ModuleReference/Text/index.jsx:
```diff
@@ -0,0 +1,31 @@
+        <DocsLayout>
+            <Helmet title="Text Modules | Streamr Docs" />
+            {(// $FlowFixMe
+                Object.entries(canvasModulesCategorised).map(([category, helps]: [string, any[]]) => (
```
Does this need to do any looping at all if it's just looping to find a single entry? i.e.

```JS
export default function TextModules() {
    const category = 'Text'
    const helps = canvasModulesCategorised[category]
    return (
        <DocsLayout>
            <Helmet title="Text Modules | Streamr Docs" />
            <section key={category} className={docsStyles.canvasModule}>
                <h1 className={docsStyles.mdH1}>{category}</h1>
                    {helps.map((m) => (
                        <CanvasModuleHelp key={m.id} module={m} help={m.help} />
                     ))}
            </section>
        </DocsLayout>
    )
}
```

And then if `category` was a prop, all of these individual category listing components could use a single component:

```JS
export default function HelpModules({ category, pageTitle }) {
    const helps = canvasModulesCategorised[category]
    return (
        <DocsLayout>
            <DocsHelmet title={pageTitle} />
            <section key={category} className={docsStyles.canvasModule}>
                <h1 className={docsStyles.mdH1}>{category}</h1>
                    {helps.map((m) => (
                        <CanvasModuleHelp key={m.id} module={m} help={m.help} />
                     ))}
            </section>
        </DocsLayout>
    )
}
```

fonty1 replied:
Good point! I didn't look too much into this at the time, this approach seems like a time saving refactor.

timoxley commented on app/src/docs/components/PageTurner/index.jsx:
```diff
@@ -3,82 +3,108 @@
     location: Location,
 }
 
 class PageTurner extends React.Component<Props> {
```
Not sure but I wonder if the built-in utilities in `react-router` could be of assistance here:  e.g. https://reacttraining.com/react-router/web/api/matchPath

fonty1 replied:
I didn't think of those at the time, after a quick glance I'd say, *probably* but I don't think there's much value in this particular refactor.

fonty1 commented:
Thanks @timoxley for the help and fresh eyes, implemented:
- Navigation into it's own component `TableOfContents`
- `moduleReference` pages to a single component with route render props of `pageTitle` and `category`
- usage of `DocsHelmet` component w/ title template
- various object/array filtering refactors

fonty1 commented:
@mattatgit this is working for me now, I hope it works for you too!

http://streamr-marketplace-pr-c698b215ca14811ed284e050c60c0bfd88e2743f.s3-website-eu-west-1.amazonaws.com/

fonty1 commented:
@mattatgit please take another look at the new URL: 
http://streamr-marketplace-pr-825330103b1394f929806d2250bb87cc3bbe267c.s3-website-eu-west-1.amazonaws.com/docs/introduction

*Change log:*
- Shortened tutorial sub-nav titles
- Added navigation top padding on scroll
- Action buttons `4px` -> `8px` of horizontal padding
- Navigation shifted to the left on high res (>1200px)
- Tightened line break to occur in the navigation at a horizontal width of 150px
- Scroll always visible on wide code snippets
- Fixed problematic 'using canvases' desktop imgs

mattatgit commented:
@fonty1 Looks good to me. Haven't checked mobile & tablet though, just desktop.

mattatgit commented:
@fonty1 One thing is that some of this stuff eg Intro to Products contains work that hasn't been deployed yet as its for the community products stuff. So not sure if that should be made public yet.

fonty1 commented:
> @fonty1 Looks good to me. Haven't checked mobile & tablet though, just desktop.

OK, I'll close the recently opened tickets addressed in the latest commit and wait on the mobile/tablet design feedback pass.

> @fonty1 One thing is that some of this stuff eg Intro to Products contains work that hasn't been deployed yet as its for the community products stuff. So not sure if that should be made public yet.

I won't merge this without your approval. I can also hide certain sections and merge the rest, up to you.

juhah commented on app/package.json:
```diff
@@ -212,7 +212,6 @@
     "react-responsive": "^4.1.0",
     "react-router-dom": "^5.0.1",
     "react-scrollable-anchor": "^0.6.1",
-    "react-scrollspy": "^3.4.0",
```
@fonty1 I merged #712 that's using `react-scrollspy` so please leave this in :)

fonty1 replied:
😛I shouldn't have touched it in the first place, good little library!

fonty1 commented:
@mattatgit Mob nav fix: Docs-92, Docs-93 Mob line breaks MP page fix, Docs-94 Highlight type size fix (warning component) should be fixed - viewable here:

http://streamr-marketplace-pr-4b0c82e5b7ea5e81e55c3c5b08a6fcb1a938a325.s3-website-eu-west-1.amazonaws.com/docs/streams/partitioning

fonty1 commented:
@mattatgit please see:
http://streamr-marketplace-pr-2a7b0429906f4bfae87a3fc1cd30d019ddf49157.s3-website-eu-west-1.amazonaws.com/docs/module-reference/boolean

Updated designs for:
- Module reference docs
- Editor sidebar module help

mattatgit commented:
@fonty1 Some pretty good progress there. 

Makes it very clear which module explanation .md files still need to be cleaned up. I think Tim made me a branch in which to stack up edits to the .md files which can then be merged in one go rather than committing each .md file.

Also there seem to be a number of the port explanations that are not using sentence case but need to be ie they look like this: 

**PortName** String. an explanation without sentence case.

I can do a pass on the .md explanations, but maybe you can work on fixing that problem ?

I think there are maybe some styling improvements to make, but even so this is definitely an improvement over what we have.

mattatgit commented:
Just looking at the sidebar module docs now, and i think they do need to have some more subheads, so will add those in.

timoxley commented:
@fonty1 CI currently failing due to broken `broker-node`, unrelated to this PR. Should probably merge this as-is and continue with ongoing changes in new PRs.

fonty1 commented:
> @fonty1 CI currently failing due to broken `broker-node`, unrelated to this PR. Should probably merge this as-is and continue with ongoing changes in new PRs.

Fair enough, I suppose it's pretty much ready for prime time anyway 👍
